In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [ ]:
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from losses.dicece_loss import DiceCELoss
from models.nnunet import nnUNet3D
from utils.unets_helper_functions import (
    set_seed,
    train_one_epoch,
    validate_one_epoch,
    save_checkpoint,
    PatchDataset,
    evaluate_full_volume
)

In [3]:
set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)


Using device: cuda


In [4]:
FOLD = 0

with open(f"../data/splits/fold_{FOLD}/train.txt") as f:
    train_cases = f.read().splitlines()

with open(f"../data/splits/fold_{FOLD}/val.txt") as f:
    val_cases = f.read().splitlines()

print("Train cases:", len(train_cases))
print("Val cases:", len(val_cases))


Train cases: 33
Val cases: 9


- Patch size = 96³
- Patches per case = 12
- Foreground sampling = 0.6
- Batch size = 2 (fallback 1 if OOM)

- Loss = Dice + CE
- Weights = [0.05,1,1,1,1,1,1]
- Optimizer = AdamW (lr 2e-4)
- Scheduler = CosineAnnealing
- Epochs = 80–100


In [5]:
PATCH_SIZE = 96
PATCHES_PER_CASE = 12
FOREGROUND_PROB = 0.6
NUM_WORKERS = 2
train_dataset = PatchDataset(train_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = PATCHES_PER_CASE,
                        patch_size = PATCH_SIZE,
                        foreground_prob = FOREGROUND_PROB,
                        augment=True)

val_dataset = PatchDataset(val_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = 1,
                        patch_size = PATCH_SIZE,
                        foreground_prob = FOREGROUND_PROB,
                        augment=False)


train_loader = DataLoader(
    train_dataset,
    batch_size=2,
    shuffle=True,
    num_workers=NUM_WORKERS,      
    pin_memory=True    
)

val_loader = DataLoader(
    val_dataset,
    batch_size=1,
    num_workers=NUM_WORKERS,    
    pin_memory=True
)

In [6]:
model = nnUNet3D(
    in_channels=1,
    out_channels=7,
    base_filters=24   # or try 32
).to(device)

print("nnU-Net style model Initialized")


nnU-Net style model Initialized


In [7]:
EPOCHS = 100

optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4)

scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=EPOCHS
)

weights = torch.tensor([0.05,1,1,1,1,1,1]).to(device)
loss_fn = DiceCELoss(weight=weights)
scaler = torch.cuda.amp.GradScaler()


C:\Users\dhanu\AppData\Local\Temp\ipykernel_18800\500315588.py:11: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [ ]:
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "final_nnunet_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)
best_val_loss = float("inf")

best_val_loss = float("inf")

for epoch in range(EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        loss_fn,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        loss_fn,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

    # Save checkpoint every 10 epochs
    if (epoch + 1) % 10 == 0:
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, f"checkpoint_epoch_{epoch+1}.pth")
        )
        print("Checkpoint Saved")
    
    scheduler.step()

100%|██████████| 198/198 [11:40<00:00,  3.54s/it]



Epoch 0
Train Loss: 3.0295
Val Loss: 2.4004
Per Class Dice: [0.24096159 0.01826135 0.11111149 0.04302141 0.14297028 0.31391542]
Saved Best Model


100%|██████████| 198/198 [12:40<00:00,  3.84s/it]



Epoch 1
Train Loss: 2.1671
Val Loss: 1.8755
Per Class Dice: [0.62953296 0.41648374 0.3234607  0.32071477 0.39863618 0.6882978 ]
Saved Best Model


100%|██████████| 198/198 [10:33<00:00,  3.20s/it]



Epoch 2
Train Loss: 1.7921
Val Loss: 1.3477
Per Class Dice: [0.58110946 0.65480369 0.65281436 0.30977599 0.44635693 0.65476642]
Saved Best Model


100%|██████████| 198/198 [10:48<00:00,  3.28s/it]



Epoch 3
Train Loss: 1.3491
Val Loss: 1.3474
Per Class Dice: [0.55665338 0.5716234  0.58076863 0.24179151 0.12668036 0.63083108]
Saved Best Model


100%|██████████| 198/198 [11:41<00:00,  3.54s/it]



Epoch 4
Train Loss: 1.2995
Val Loss: 0.4583
Per Class Dice: [0.88952948 0.93443843 0.8128129  0.76641907 0.75438415 0.89827537]
Saved Best Model


100%|██████████| 198/198 [14:01<00:00,  4.25s/it]



Epoch 5
Train Loss: 1.1987
Val Loss: 1.5199
Per Class Dice: [0.68702805 0.65625074 0.55537091 0.37627625 0.33225539 0.68661239]


100%|██████████| 198/198 [13:04<00:00,  3.96s/it]



Epoch 6
Train Loss: 1.1513
Val Loss: 0.7978
Per Class Dice: [0.86695794 0.85841524 0.83550746 0.41853004 0.5316298  0.84332483]


100%|██████████| 198/198 [10:20<00:00,  3.13s/it]



Epoch 7
Train Loss: 1.2197
Val Loss: 1.2937
Per Class Dice: [0.54683546 0.69905771 0.71364007 0.32554004 0.43225274 0.53138543]


100%|██████████| 198/198 [10:32<00:00,  3.19s/it]



Epoch 8
Train Loss: 1.1480
Val Loss: 1.7489
Per Class Dice: [0.58865503 0.90757278 0.84287051 0.62181485 0.52240194 0.82019388]


100%|██████████| 198/198 [12:33<00:00,  3.81s/it]



Epoch 9
Train Loss: 1.1431
Val Loss: 0.7884
Per Class Dice: [0.82563025 0.67863489 0.69277054 0.4701305  0.59702941 0.7894345 ]
Checkpoint Saved


100%|██████████| 198/198 [13:06<00:00,  3.97s/it]



Epoch 10
Train Loss: 1.1111
Val Loss: 1.1859
Per Class Dice: [0.56467845 0.64258723 0.6938004  0.33440292 0.41196631 0.6753026 ]


100%|██████████| 198/198 [13:21<00:00,  4.05s/it]



Epoch 11
Train Loss: 1.0950
Val Loss: 1.3636
Per Class Dice: [0.53528094 0.76390115 0.61575485 0.51867371 0.16697135 0.55776146]


100%|██████████| 198/198 [14:27<00:00,  4.38s/it]



Epoch 12
Train Loss: 1.1118
Val Loss: 0.7016
Per Class Dice: [0.90191427 0.7802958  0.83535484 0.42181457 0.62475289 0.86093298]


100%|██████████| 198/198 [13:07<00:00,  3.98s/it]



Epoch 13
Train Loss: 1.1177
Val Loss: 0.8034
Per Class Dice: [0.80789496 0.84906737 0.75776477 0.58354233 0.64285136 0.76479534]


100%|██████████| 198/198 [13:36<00:00,  4.12s/it]



Epoch 14
Train Loss: 1.0485
Val Loss: 0.8707
Per Class Dice: [0.7045369  0.89079375 0.87463135 0.5297751  0.73029787 0.62488609]


100%|██████████| 198/198 [10:32<00:00,  3.19s/it]



Epoch 15
Train Loss: 1.0301
Val Loss: 1.4081
Per Class Dice: [0.70839737 0.40015115 0.58883661 0.23677342 0.4201562  0.44477187]


100%|██████████| 198/198 [10:34<00:00,  3.21s/it]



Epoch 16
Train Loss: 1.0637
Val Loss: 1.4430
Per Class Dice: [0.34156656 0.77302664 0.85645682 0.19286953 0.35721829 0.69444437]


100%|██████████| 198/198 [10:44<00:00,  3.25s/it]



Epoch 17
Train Loss: 1.0626
Val Loss: 1.0446
Per Class Dice: [0.58450132 0.81162983 0.87815972 0.53048529 0.30784671 0.89585888]


100%|██████████| 198/198 [10:37<00:00,  3.22s/it]



Epoch 18
Train Loss: 1.1111
Val Loss: 0.8934
Per Class Dice: [0.80969284 0.69310749 0.69450668 0.44615619 0.371862   0.76494027]


100%|██████████| 198/198 [10:03<00:00,  3.05s/it]



Epoch 19
Train Loss: 0.9753
Val Loss: 0.8537
Per Class Dice: [0.71581004 0.93663424 0.92546659 0.57706022 0.5522804  0.80287563]
Checkpoint Saved


100%|██████████| 198/198 [10:28<00:00,  3.17s/it]



Epoch 20
Train Loss: 1.0627
Val Loss: 1.2623
Per Class Dice: [0.66926857 0.83262446 0.54711199 0.16784736 0.34568017 0.61359021]


100%|██████████| 198/198 [10:11<00:00,  3.09s/it]



Epoch 21
Train Loss: 0.9738
Val Loss: 0.9719
Per Class Dice: [0.70621016 0.84195884 0.85749894 0.48258206 0.57495084 0.92368066]


100%|██████████| 198/198 [10:28<00:00,  3.17s/it]



Epoch 22
Train Loss: 0.9988
Val Loss: 0.8341
Per Class Dice: [0.8897773  0.78727341 0.78381491 0.15224966 0.57793944 0.83624997]


100%|██████████| 198/198 [10:55<00:00,  3.31s/it]



Epoch 23
Train Loss: 1.0366
Val Loss: 0.9742
Per Class Dice: [0.9138728  0.69844421 0.69504203 0.32952612 0.59515312 0.79407168]


100%|██████████| 198/198 [10:20<00:00,  3.14s/it]



Epoch 24
Train Loss: 1.0082
Val Loss: 0.9590
Per Class Dice: [0.79063995 0.87881537 0.86958805 0.46658226 0.41822598 0.89524941]


100%|██████████| 198/198 [10:33<00:00,  3.20s/it]



Epoch 25
Train Loss: 1.0160
Val Loss: 1.0249
Per Class Dice: [0.70337423 0.807265   0.87905298 0.59294074 0.47247158 0.84680117]


100%|██████████| 198/198 [11:06<00:00,  3.37s/it]



Epoch 26
Train Loss: 0.9830
Val Loss: 1.1962
Per Class Dice: [0.61597679 0.9247848  0.66852761 0.49410464 0.62915094 0.81252956]


100%|██████████| 198/198 [10:52<00:00,  3.30s/it]



Epoch 27
Train Loss: 0.9351
Val Loss: 0.9220
Per Class Dice: [0.86057456 0.75550156 0.7189845  0.33241072 0.26155913 0.81031603]


100%|██████████| 198/198 [10:27<00:00,  3.17s/it]



Epoch 28
Train Loss: 1.0248
Val Loss: 1.0577
Per Class Dice: [0.72830637 0.68493513 0.84667031 0.48116565 0.38592341 0.55934497]


100%|██████████| 198/198 [10:48<00:00,  3.27s/it]



Epoch 29
Train Loss: 0.9681
Val Loss: 1.3376
Per Class Dice: [0.56405743 0.63294373 0.716186   0.08890943 0.25153283 0.76219232]
Checkpoint Saved


100%|██████████| 198/198 [10:51<00:00,  3.29s/it]



Epoch 30
Train Loss: 0.9390
Val Loss: 0.7756
Per Class Dice: [0.60494994 0.75445943 0.78267607 0.64997805 0.71803687 0.79882668]


100%|██████████| 198/198 [10:14<00:00,  3.10s/it]



Epoch 31
Train Loss: 0.9897
Val Loss: 0.8422
Per Class Dice: [0.82195599 0.66486895 0.8543081  0.42295254 0.55538623 0.7889911 ]


100%|██████████| 198/198 [10:35<00:00,  3.21s/it]



Epoch 32
Train Loss: 1.0155
Val Loss: 0.8935
Per Class Dice: [0.74894797 0.78871366 0.91793294 0.60962498 0.49574567 0.68734073]


100%|██████████| 198/198 [09:54<00:00,  3.00s/it]



Epoch 33
Train Loss: 0.9819
Val Loss: 1.1490
Per Class Dice: [0.71235486 0.45346393 0.73290836 0.18988644 0.33176391 0.75883902]


100%|██████████| 198/198 [10:21<00:00,  3.14s/it]



Epoch 34
Train Loss: 0.9450
Val Loss: 0.7289
Per Class Dice: [0.9292765  0.81977922 0.91479004 0.42466166 0.46368717 0.78576908]


100%|██████████| 198/198 [10:51<00:00,  3.29s/it]



Epoch 35
Train Loss: 1.0138
Val Loss: 0.6857
Per Class Dice: [0.91889018 0.83461368 0.8223297  0.42708021 0.51781499 0.85186326]


100%|██████████| 198/198 [11:02<00:00,  3.35s/it]



Epoch 36
Train Loss: 0.9703
Val Loss: 1.1656
Per Class Dice: [0.68815962 0.71296108 0.75522577 0.46327407 0.23210329 0.62170617]


100%|██████████| 198/198 [10:17<00:00,  3.12s/it]



Epoch 37
Train Loss: 0.9289
Val Loss: 0.6060
Per Class Dice: [0.89379433 0.86863401 0.86142991 0.62578157 0.49885862 0.81393549]


100%|██████████| 198/198 [10:10<00:00,  3.08s/it]



Epoch 38
Train Loss: 0.9518
Val Loss: 0.9800
Per Class Dice: [0.68091751 0.86960428 0.78197675 0.25131904 0.42911618 0.66382244]


100%|██████████| 198/198 [10:09<00:00,  3.08s/it]



Epoch 39
Train Loss: 0.9341
Val Loss: 0.8002
Per Class Dice: [0.88814361 0.78285949 0.64599792 0.30465256 0.55404312 0.83455332]
Checkpoint Saved


100%|██████████| 198/198 [10:33<00:00,  3.20s/it]



Epoch 40
Train Loss: 0.8936
Val Loss: 0.9107
Per Class Dice: [0.80801435 0.91420458 0.79803299 0.3305158  0.42831805 0.79630042]


100%|██████████| 198/198 [10:33<00:00,  3.20s/it]



Epoch 41
Train Loss: 0.9592
Val Loss: 0.6408
Per Class Dice: [0.92172919 0.78645186 0.90302367 0.479805   0.60040889 0.85851293]


100%|██████████| 198/198 [10:27<00:00,  3.17s/it]



Epoch 42
Train Loss: 0.9452
Val Loss: 0.9196
Per Class Dice: [0.81883652 0.92925472 0.80002606 0.37364625 0.32682447 0.88594353]


100%|██████████| 198/198 [10:54<00:00,  3.31s/it]



Epoch 43
Train Loss: 0.8821
Val Loss: 1.0605
Per Class Dice: [0.60858692 0.79921422 0.80942308 0.41310825 0.5190441  0.72168744]


100%|██████████| 198/198 [10:01<00:00,  3.04s/it]



Epoch 44
Train Loss: 0.8830
Val Loss: 0.8171
Per Class Dice: [0.82598835 0.81912384 0.85242441 0.36553283 0.47687356 0.89362595]


100%|██████████| 198/198 [10:31<00:00,  3.19s/it]



Epoch 45
Train Loss: 0.8938
Val Loss: 0.8482
Per Class Dice: [0.90281451 0.77314209 0.83109686 0.28989992 0.63293142 0.71342099]


100%|██████████| 198/198 [11:02<00:00,  3.34s/it]



Epoch 46
Train Loss: 0.9211
Val Loss: 0.5382
Per Class Dice: [0.84211541 0.93036295 0.87841804 0.67707598 0.64945167 0.76082578]


 61%|██████    | 120/198 [06:02<02:24,  1.85s/it]

In [8]:
#  RESUME TRAINING 
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "final_nnunet_fold0")

start_epoch = 0
best_val_loss = float("inf")

resume_path = os.path.join(SAVE_DIR, "checkpoint_epoch_40.pth")  

if os.path.exists(resume_path):
    print("Loading checkpoint:", resume_path)

    checkpoint = torch.load(resume_path, map_location=device)

    model.load_state_dict(checkpoint["model_state_dict"])
    optimizer.load_state_dict(checkpoint["optimizer_state_dict"])

    start_epoch = checkpoint["epoch"] + 1
    best_val_loss = checkpoint["best_val_loss"]
    scheduler.last_epoch = start_epoch - 1

    print(f"Resuming from epoch {start_epoch}")
else:
    print("No checkpoint found, starting fresh.")


Loading checkpoint: c:\Users\dhanu\Desktop\mini-project-1\experiments\final_nnunet_fold0\checkpoint_epoch_40.pth


C:\Users\dhanu\AppData\Local\Temp\ipykernel_18800\833551612.py:13: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(resume_path, map_location=device)


Resuming from epoch 40


In [9]:
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "final_nnunet_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)
# best_val_loss = float("inf")

for epoch in range(start_epoch,EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        loss_fn,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        loss_fn,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

    # Save checkpoint every 10 epochs
    if (epoch + 1) % 10 == 0:
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, f"checkpoint_epoch_{epoch+1}.pth")
        )
        print("Checkpoint Saved")
    
    scheduler.step()

100%|██████████| 198/198 [13:08<00:00,  3.98s/it]



Epoch 40
Train Loss: 0.9182
Val Loss: 1.0648
Per Class Dice: [0.78964309 0.68708734 0.78297799 0.4438036  0.3625842  0.71850369]


100%|██████████| 198/198 [13:53<00:00,  4.21s/it]



Epoch 41
Train Loss: 0.9196
Val Loss: 0.7962
Per Class Dice: [0.93717047 0.73280533 0.79892306 0.27778749 0.5064058  0.85469522]


100%|██████████| 198/198 [12:59<00:00,  3.93s/it]



Epoch 42
Train Loss: 0.9279
Val Loss: 0.7662
Per Class Dice: [0.82477357 0.85282486 0.88546491 0.38132631 0.48408107 0.73329212]


100%|██████████| 198/198 [13:18<00:00,  4.03s/it]



Epoch 43
Train Loss: 0.8708
Val Loss: 0.9333
Per Class Dice: [0.90864041 0.87840722 0.79981778 0.46394558 0.22880004 0.83121306]


100%|██████████| 198/198 [13:09<00:00,  3.99s/it]



Epoch 44
Train Loss: 0.9033
Val Loss: 0.3212
Per Class Dice: [0.94699619 0.94717067 0.86108263 0.81151613 0.73856206 0.93819277]
Saved Best Model


100%|██████████| 198/198 [12:55<00:00,  3.91s/it]



Epoch 45
Train Loss: 0.8672
Val Loss: 0.8713
Per Class Dice: [0.85465487 0.91310751 0.89433925 0.31519551 0.34145057 0.73357998]


100%|██████████| 198/198 [13:19<00:00,  4.04s/it]



Epoch 46
Train Loss: 0.8809
Val Loss: 0.6553
Per Class Dice: [0.91638347 0.87594274 0.85713635 0.55479917 0.59318122 0.86640754]


100%|██████████| 198/198 [13:12<00:00,  4.00s/it]



Epoch 47
Train Loss: 0.9327
Val Loss: 0.9188
Per Class Dice: [0.76078113 0.8408557  0.86156923 0.37734182 0.36805394 0.66136375]


100%|██████████| 198/198 [13:46<00:00,  4.18s/it]



Epoch 48
Train Loss: 0.8681
Val Loss: 0.6694
Per Class Dice: [0.8160732  0.93224183 0.82199412 0.52137376 0.52680175 0.89678546]


100%|██████████| 198/198 [12:20<00:00,  3.74s/it]



Epoch 49
Train Loss: 0.8737
Val Loss: 0.6319
Per Class Dice: [0.84450238 0.80288255 0.81022323 0.49228909 0.61766077 0.90405215]
Checkpoint Saved


100%|██████████| 198/198 [10:04<00:00,  3.05s/it]



Epoch 50
Train Loss: 0.8426
Val Loss: 1.0035
Per Class Dice: [0.54109916 0.85895778 0.83613518 0.45948338 0.56394651 0.69079427]


100%|██████████| 198/198 [10:37<00:00,  3.22s/it]



Epoch 51
Train Loss: 0.8870
Val Loss: 1.1964
Per Class Dice: [0.83854365 0.79399393 0.88834274 0.51302617 0.20072313 0.73408816]


100%|██████████| 198/198 [10:46<00:00,  3.26s/it]



Epoch 52
Train Loss: 0.8696
Val Loss: 0.6653
Per Class Dice: [0.92260088 0.83685468 0.85528053 0.43879034 0.636557   0.88078751]


100%|██████████| 198/198 [10:16<00:00,  3.11s/it]



Epoch 53
Train Loss: 0.8917
Val Loss: 0.5833
Per Class Dice: [0.92965809 0.88901352 0.89081436 0.48840459 0.68550451 0.88811286]


100%|██████████| 198/198 [10:12<00:00,  3.09s/it]



Epoch 54
Train Loss: 0.8295
Val Loss: 0.5296
Per Class Dice: [0.83129071 0.89555981 0.87660468 0.62260301 0.7092454  0.88804873]


100%|██████████| 198/198 [10:18<00:00,  3.13s/it]



Epoch 55
Train Loss: 0.8337
Val Loss: 1.1958
Per Class Dice: [0.60387041 0.85489498 0.69174492 0.17604101 0.35612629 0.5560754 ]


100%|██████████| 198/198 [10:53<00:00,  3.30s/it]



Epoch 56
Train Loss: 0.8597
Val Loss: 0.7289
Per Class Dice: [0.80711402 0.89442717 0.74221844 0.38343344 0.61731958 0.93052936]


100%|██████████| 198/198 [10:36<00:00,  3.21s/it]



Epoch 57
Train Loss: 0.7991
Val Loss: 0.9724
Per Class Dice: [0.61407143 0.70816725 0.87590176 0.53328994 0.41753749 0.79216856]


100%|██████████| 198/198 [10:28<00:00,  3.18s/it]



Epoch 58
Train Loss: 0.8632
Val Loss: 0.8239
Per Class Dice: [0.80919142 0.73566143 0.82671014 0.40151332 0.37795253 0.76600265]


100%|██████████| 198/198 [09:55<00:00,  3.01s/it]



Epoch 59
Train Loss: 0.8115
Val Loss: 0.5403
Per Class Dice: [0.89247909 0.82631584 0.92491251 0.55353296 0.60234045 0.9031284 ]
Checkpoint Saved


100%|██████████| 198/198 [10:28<00:00,  3.18s/it]



Epoch 60
Train Loss: 0.8031
Val Loss: 0.9846
Per Class Dice: [0.92452564 0.82627381 0.67701445 0.44310285 0.5638405  0.84167998]


100%|██████████| 198/198 [10:16<00:00,  3.11s/it]



Epoch 61
Train Loss: 0.7973
Val Loss: 0.9804
Per Class Dice: [0.83152233 0.84134926 0.87375733 0.58976107 0.55301    0.92579118]


100%|██████████| 198/198 [10:29<00:00,  3.18s/it]



Epoch 62
Train Loss: 0.8124
Val Loss: 0.8212
Per Class Dice: [0.90293033 0.77205431 0.79100509 0.22368072 0.57239136 0.83682845]


100%|██████████| 198/198 [10:56<00:00,  3.32s/it]



Epoch 63
Train Loss: 0.8155
Val Loss: 0.8209
Per Class Dice: [0.90495819 0.72880177 0.85263958 0.41502114 0.52428881 0.82754633]


100%|██████████| 198/198 [10:52<00:00,  3.30s/it]



Epoch 64
Train Loss: 0.7812
Val Loss: 0.7804
Per Class Dice: [0.80610184 0.88780452 0.89005091 0.46701417 0.53304175 0.90013641]


100%|██████████| 198/198 [10:22<00:00,  3.15s/it]



Epoch 65
Train Loss: 0.8090
Val Loss: 0.7570
Per Class Dice: [0.67330455 0.81232495 0.89332432 0.44538951 0.51290879 0.86678388]


100%|██████████| 198/198 [10:57<00:00,  3.32s/it]



Epoch 66
Train Loss: 0.8003
Val Loss: 0.6311
Per Class Dice: [0.94737812 0.93651209 0.70228002 0.61819576 0.72031469 0.80309382]


100%|██████████| 198/198 [10:44<00:00,  3.25s/it]



Epoch 67
Train Loss: 0.7412
Val Loss: 0.8132
Per Class Dice: [0.90188532 0.80546669 0.82247264 0.344505   0.39115388 0.84220552]


100%|██████████| 198/198 [10:34<00:00,  3.20s/it]



Epoch 68
Train Loss: 0.7665
Val Loss: 0.7712
Per Class Dice: [0.80985039 0.91914028 0.87105044 0.45189543 0.49080259 0.87827898]


100%|██████████| 198/198 [10:38<00:00,  3.23s/it]



Epoch 69
Train Loss: 0.7791
Val Loss: 0.7920
Per Class Dice: [0.82859838 0.8557811  0.87076657 0.31278543 0.42918467 0.8312602 ]
Checkpoint Saved


100%|██████████| 198/198 [10:52<00:00,  3.30s/it]



Epoch 70
Train Loss: 0.7573
Val Loss: 0.4601
Per Class Dice: [0.83760937 0.89256871 0.89672342 0.67533733 0.70851519 0.90932808]


100%|██████████| 198/198 [10:25<00:00,  3.16s/it]



Epoch 71
Train Loss: 0.7692
Val Loss: 0.5343
Per Class Dice: [0.93098468 0.92970773 0.85237608 0.5505658  0.67535431 0.90827166]


100%|██████████| 198/198 [10:40<00:00,  3.23s/it]



Epoch 72
Train Loss: 0.8056
Val Loss: 0.4265
Per Class Dice: [0.93020093 0.92320186 0.93056254 0.66855611 0.64912764 0.87986142]


100%|██████████| 198/198 [10:04<00:00,  3.05s/it]



Epoch 73
Train Loss: 0.7371
Val Loss: 0.9778
Per Class Dice: [0.82066056 0.7592496  0.78329706 0.33202559 0.44501616 0.78037089]


100%|██████████| 198/198 [10:25<00:00,  3.16s/it]



Epoch 74
Train Loss: 0.7725
Val Loss: 0.6853
Per Class Dice: [0.92798316 0.8181999  0.91652865 0.50138521 0.46204792 0.82429342]


100%|██████████| 198/198 [10:48<00:00,  3.28s/it]



Epoch 75
Train Loss: 0.7696
Val Loss: 0.6423
Per Class Dice: [0.92689843 0.83885409 0.84464695 0.44281183 0.51459403 0.86647328]


100%|██████████| 198/198 [10:41<00:00,  3.24s/it]



Epoch 76
Train Loss: 0.7891
Val Loss: 0.8859
Per Class Dice: [0.92762846 0.84268402 0.87753161 0.67246674 0.35880102 0.71816716]


100%|██████████| 198/198 [10:13<00:00,  3.10s/it]



Epoch 77
Train Loss: 0.7490
Val Loss: 0.4930
Per Class Dice: [0.94411316 0.87464821 0.88262486 0.65969026 0.60601388 0.90113158]


100%|██████████| 198/198 [10:08<00:00,  3.07s/it]



Epoch 78
Train Loss: 0.7212
Val Loss: 0.7289
Per Class Dice: [0.94549729 0.87361801 0.81785886 0.51768156 0.6736269  0.91455936]


100%|██████████| 198/198 [10:09<00:00,  3.08s/it]



Epoch 79
Train Loss: 0.7282
Val Loss: 0.7448
Per Class Dice: [0.91591772 0.82942659 0.7822453  0.28383664 0.56212778 0.83528853]
Checkpoint Saved


100%|██████████| 198/198 [10:29<00:00,  3.18s/it]



Epoch 80
Train Loss: 0.7437
Val Loss: 0.7374
Per Class Dice: [0.89870891 0.89071639 0.91278317 0.35535288 0.43677997 0.85220363]


100%|██████████| 198/198 [10:29<00:00,  3.18s/it]



Epoch 81
Train Loss: 0.7673
Val Loss: 0.5819
Per Class Dice: [0.92015298 0.79593756 0.90718623 0.46590273 0.71284418 0.86183507]


100%|██████████| 198/198 [10:18<00:00,  3.13s/it]



Epoch 82
Train Loss: 0.7321
Val Loss: 0.6382
Per Class Dice: [0.94413457 0.9297062  0.81162326 0.41783996 0.56492074 0.87514739]


100%|██████████| 198/198 [10:53<00:00,  3.30s/it]



Epoch 83
Train Loss: 0.7416
Val Loss: 0.4823
Per Class Dice: [0.89520127 0.87967371 0.93779395 0.65599468 0.74761932 0.82351366]


100%|██████████| 198/198 [10:07<00:00,  3.07s/it]



Epoch 84
Train Loss: 0.7367
Val Loss: 0.6408
Per Class Dice: [0.93655209 0.85809646 0.86516702 0.40335374 0.61231061 0.89298891]


100%|██████████| 198/198 [10:33<00:00,  3.20s/it]



Epoch 85
Train Loss: 0.7262
Val Loss: 0.7281
Per Class Dice: [0.80893532 0.84324723 0.86469933 0.31821676 0.63369922 0.85894678]


100%|██████████| 198/198 [16:21<00:00,  4.96s/it] 



Epoch 86
Train Loss: 0.7573
Val Loss: 0.4026
Per Class Dice: [0.95389644 0.93313193 0.89047998 0.63880113 0.78030094 0.90313109]


100%|██████████| 198/198 [10:22<00:00,  3.14s/it]



Epoch 87
Train Loss: 0.6965
Val Loss: 0.6495
Per Class Dice: [0.72395715 0.86657951 0.91926561 0.50043635 0.69147204 0.93356633]


100%|██████████| 198/198 [12:31<00:00,  3.80s/it]



Epoch 88
Train Loss: 0.7476
Val Loss: 0.4800
Per Class Dice: [0.93642141 0.90790572 0.90289829 0.63136361 0.68671016 0.87386205]


100%|██████████| 198/198 [13:29<00:00,  4.09s/it]



Epoch 89
Train Loss: 0.7231
Val Loss: 0.6112
Per Class Dice: [0.93071302 0.88432924 0.8544606  0.55919921 0.43104876 0.86954183]
Checkpoint Saved


100%|██████████| 198/198 [12:23<00:00,  3.76s/it]



Epoch 90
Train Loss: 0.6902
Val Loss: 0.7202
Per Class Dice: [0.90929593 0.86499592 0.84525186 0.44715035 0.53660625 0.85098945]


100%|██████████| 198/198 [12:04<00:00,  3.66s/it]



Epoch 91
Train Loss: 0.7490
Val Loss: 0.5229
Per Class Dice: [0.93619968 0.90009054 0.87687131 0.59254122 0.6610041  0.89864718]


100%|██████████| 198/198 [09:46<00:00,  2.96s/it]



Epoch 92
Train Loss: 0.7004
Val Loss: 0.5709
Per Class Dice: [0.73438459 0.90553576 0.92277397 0.74272822 0.53161573 0.80995094]


100%|██████████| 198/198 [09:48<00:00,  2.97s/it]



Epoch 93
Train Loss: 0.7584
Val Loss: 0.8217
Per Class Dice: [0.89488803 0.84384506 0.76701073 0.26823229 0.56570132 0.86235683]


100%|██████████| 198/198 [09:55<00:00,  3.01s/it]



Epoch 94
Train Loss: 0.7241
Val Loss: 0.6101
Per Class Dice: [0.84191346 0.90109636 0.87179002 0.66451811 0.61670377 0.82694894]


100%|██████████| 198/198 [10:17<00:00,  3.12s/it]



Epoch 95
Train Loss: 0.7339
Val Loss: 0.5507
Per Class Dice: [0.93263782 0.89491368 0.76880535 0.67748445 0.58443943 0.87719736]


100%|██████████| 198/198 [09:46<00:00,  2.96s/it]



Epoch 96
Train Loss: 0.6852
Val Loss: 0.9440
Per Class Dice: [0.84835183 0.66591193 0.92065999 0.28956462 0.5158661  0.77973872]


100%|██████████| 198/198 [10:02<00:00,  3.04s/it]



Epoch 97
Train Loss: 0.7020
Val Loss: 0.6172
Per Class Dice: [0.84450788 0.90648358 0.83997999 0.48274405 0.71241564 0.86756296]


100%|██████████| 198/198 [10:23<00:00,  3.15s/it]



Epoch 98
Train Loss: 0.7678
Val Loss: 0.6323
Per Class Dice: [0.94198663 0.86449186 0.88054056 0.62228538 0.60711136 0.92772843]


100%|██████████| 198/198 [10:37<00:00,  3.22s/it]



Epoch 99
Train Loss: 0.7100
Val Loss: 0.4747
Per Class Dice: [0.94630977 0.90762073 0.88696324 0.57360429 0.64115595 0.90347299]
Checkpoint Saved


In [5]:
print("Training Complete")

Training Complete


In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = nnUNet3D().to(device)

model_path = "../experiments/final_nnunet_fold0/best_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


C:\Users\dhanu\AppData\Local\Temp\ipykernel_12472\1838595719.py:6: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


Model loaded for testing.


with stride = 48

In [6]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.8978452136615971), np.float64(0.8305425319412498), np.float64(0.7538966370920906), np.float64(0.46985142283306586), np.float64(1.943256894785743e-09), np.float64(0.8762388636666358)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9025668513183946), np.float64(0.7725724022380133), np.float64(0.699588477984386), np.float64(0.41779829828997), np.float64(0.11425452641294048), np.float64(0.8680479474337746)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8408707132599949), np.float64(0.8391407429850272), np.float64(0.6774323785275891), np.float64(0.5080541239754864), np.float64(0.24437012347338896), np.float64(0.8840684206667073)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.885770474355598), np.float64(0.7636417778875592), np.float64(0.77953

stride = 48, postprocessing - off

In [6]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.8977033291462012), np.float64(0.8303571429833653), np.float64(0.7538966370920906), np.float64(0.4692425354331033), np.float64(0.1104374912167172), np.float64(0.8759881246421312)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9017134698055349), np.float64(0.7725066009943731), np.float64(0.6994892753840873), np.float64(0.4241250396720305), np.float64(0.20175729509138157), np.float64(0.8672740149582677)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8405094369063958), np.float64(0.8388254721524693), np.float64(0.6788879942004674), np.float64(0.5088580578928186), np.float64(0.2851401594643423), np.float64(0.882130557911972)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.8833870592923508), np.float64(0.7637338099123613), np.float64(0.7795309

stride = 40

In [6]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.9127681842195793), np.float64(0.8321249304181938), np.float64(0.7817444223494029), np.float64(0.5851607092331326), np.float64(0.504537205773367), np.float64(0.8775872265151359)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9071090615358719), np.float64(0.7844546596034645), np.float64(0.7147725000706104), np.float64(0.530783074737043), np.float64(0.4360771961905654), np.float64(0.8735522767136995)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8716353479330766), np.float64(0.8255271766820744), np.float64(0.7171171176834493), np.float64(0.5144499443746607), np.float64(0.23172712589959166), np.float64(0.8834769458761713)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.910668214683092), np.float64(0.7757730136731246), np.float64(0.757093723

# nnU-Net Style 3D Model — Final Results

Patch-Based Validation Performance

The final model was trained for 100 epochs using nnU-Net style patch-based training with heavy augmentation, Dice+CE loss and balanced foreground sampling.

**Best Validation Epoch
Epoch 44**

Train Loss: 0.9033
Validation Loss: 0.3212

Patch-Based Per-Class Dice (Validation)

- Mandible: 0.9470
- Brainstem: 0.9471
- Spinal Cord: 0.8611
- Parotid Left: 0.8115
- Parotid Right: 0.7386
- Oral Cavity: 0.9382

Observations (Patch Validation)

All organs achieve very high Dice (>0.80) except Parotid-R.
Strong learning of both large and thin structures.
Augmentation + DiceCE loss significantly improved generalization.
Overall, patch-based validation shows excellent learning capacity.

# Evaluating on Full Volume

(Sliding window inference — patch = 96, stride = 40, Gaussian blending + post-processing)

| Organ ID | Organ         | Mean Dice  | Std Dev |
| -------- | ------------- | ---------- | ------- |
| 1        | Bone Mandible | **0.8928** | 0.0275  |
| 2        | Brainstem     | **0.8015** | 0.0210  |
| 3        | Spinal Cord   | **0.7350** | 0.0272  |
| 4        | Parotid Left  | **0.4705** | 0.0663  |
| 5        | Parotid Right | **0.2422** | 0.1312  |
| 6        | Oral Cavity   | **0.8251** | 0.0580  |

# Full Volume Evaluation — Summary

**Strong Organs**

- Mandible (0.89) → Very stable and accurate segmentation
- Brainstem (0.80) → Consistent and reliable predictions
- Oral Cavity (0.82) → Large improvement over earlier models

These organs benefit from:
Larger anatomical context
Clear structural boundaries
Moderate Performance

- Spinal Cord (0.73)

Thin elongated structures are difficult in sliding-window inference.
Despite this, the model achieves strong performance.

Most Challenging Organs

- Parotid Left (0.47)
- Parotid Right (0.24)

Lower performance is expected due to:

Very small organ size
High anatomical variation
Sensitivity to patch sampling
Class imbalance effects

| #     | Model                             | Architecture                   | Base Filters | Patch Size | Patches / Case | FG Sampling | Augmentation           | Batch Size | Loss                   | Optimizer             | Inference                                                 |
| ----- | --------------------------------- | ------------------------------ | ------------ | ---------- | -------------- | ----------- | ---------------------- | ---------- | ---------------------- | --------------------- | --------------------------------------------------------- |
| **1** | **UNet 3D (Baseline)**            | Standard 3D U-Net              | 16           | 80³        | 6              | 0.5         |  Disabled             | 2          | Cross-Entropy          | Adam                  | Sliding window (no Gaussian)                              |
| **2** | **Attention UNet 3D**             | 3D Attention U-Net             | 16           | 80³        | 6              | 0.5         |  Disabled             | 2          | Cross-Entropy          | Adam                  | Sliding window (no Gaussian)                              |
| **3** | **UNet++ 3D**                     | Lightweight Nested U-Net       | 16           | 80³        | 6              | 0.5         |  Disabled             | 2          | Cross-Entropy          | Adam                  | Sliding window (no Gaussian)                              |
| **4** | **nnUNet-style (Initial)**        | Residual 3D U-Net              | 24           | 80³        | 6              | 0.5         |  Disabled             | 2          | Cross-Entropy          | Adam                  | Sliding window (simple averaging)                         |
| **5** | **Final nnUNet Pipeline** | Residual nnUNet-style 3D U-Net | 24           | **96³**    | **12**         | **0.6**     |  TorchIO Augmentation | 2          | **Dice + CE (Hybrid)** | **AdamW + Cosine LR** | **Sliding window + Gaussian weighting + Post-processing** |
